streams/src/test/java/org/apache/kafka/streams/processor/internals/assignment/AssignorConfigurationTest.java

```diff
@@ -54,30 +54,25 @@ public void configsShouldRejectZeroWarmups() {
     }
 
     @Test
-    public void rebalanceProtocolShouldSupportAllUpgradeFromVersions() {
+    public void shouldSupportAllUpgradeFromVersionsFromCooperativeRebalancingOn() {
+        boolean beforeCooperative = true;
         for (final UpgradeFromValues upgradeFrom : UpgradeFromValues.values()) {
-            config.put(StreamsConfig.UPGRADE_FROM_CONFIG, upgradeFrom.toString());
-            final AssignorConfiguration assignorConfiguration = new AssignorConfiguration(config);
-
-            try {
-                assignorConfiguration.rebalanceProtocol();
-            } catch (final Exception error) {
-                throw new AssertionError("Upgrade from " + upgradeFrom + " failed with " + error.getMessage() + "!");
+            if (upgradeFrom.toString().equals("2.4")) {
```

Comment:
Seems we would need to also check 2.5, ... 3.9, not just 2.4 ?

Reply:
we do, it loops over all the versions starting from the incompatible upgrades, we don't only check 2.4 it's just that 2.4 is where we switch from asserting that an exception is thrown to asserting that it can be instantiated without throwing

Reply:
Yes, we loop, but when `upgradeFrom` changes, should the condition not be:
```
if (upgradeFrom.toString().equals("2.4")
    || upgradeFrom.toString().equals("2.5")
   ...
    || upgradeFrom.toString().equals("3.9")
    // need to add a new version here for every new release    
    ) {
```
Otherwise we set `beforeCooperative = false` only for a single loop iteration when `upgradeFrom = 2.4` ?

What actually make me realize that we might want to flip from logic and initialize `beforeCooperative = false`, and set it to `true` for versions 0.10.0 to 2.3 to avoid that we need to keep updating this test for future releases.

Reply:
nope because `beforeCooperative` is declared outside the loop so once it gets updated it stays on the new value..? am I going crazy here?

I mean the testing is passing so...